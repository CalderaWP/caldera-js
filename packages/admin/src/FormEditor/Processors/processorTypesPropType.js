import PropTypes from "prop-types";

export const processorTypesPropType = PropTypes.arrayOf(PropTypes.shape({
	type: PropTypes.string,
	fields: PropTypes.array,
	conditionals: PropTypes.arrayOf(PropTypes.shape({
		rule: PropTypes.func,
		fields: PropTypes.arrayOf(PropTypes.string)
	}))
}));
