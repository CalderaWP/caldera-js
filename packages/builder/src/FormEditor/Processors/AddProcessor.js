import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Row, fieldAreaFactory} from "@calderajs/components";

const AddProcessorButtonText = ({processorTypes,value}) => {
	const findProcessorLabel = type => {
		const processor = processorTypes.find( processorType => value === processorType.type );
		if(  processor && processor.hasOwnProperty('typeLabel') ){
			return processor.typeLabel;
		}
		return type; 

	}
	if( ! value ){
		return <Fragment>Choose Processor Type</Fragment>
	}

	const label = findProcessorLabel(value);
	return <Fragment>{ `Add ${label} Processor`}</Fragment>
};

export const AddProcessor = (
	{
		setNewProcessorType,
		processorTypes,
		onCreate,
		children,
		value
	}
) => {
	const options = [];
	console.log(processorTypes)
	processorTypes.map(processorType => {
		options.push({
			value: processorType.type,
			label: processorType.hasOwnProperty( 'typeLabel') ? processorType.typeLabel : processorType.type
		});
	});

	const processorTypesField = {
		fieldType: 'select',
		label: 'Processor Type',
		fieldId: 'newProcessorType',
		required: true,
		options,
		onChange: setNewProcessorType,
		value,
	};

	
	

	return (
		<Fragment>
			{fieldAreaFactory(processorTypesField, setNewProcessorType)}
			<button 
				onClick={onCreate}
				disabled={!value}
			>
				<AddProcessorButtonText value={value} processorTypes={processorTypes}/>
			</button>
		</Fragment>
	)
};

AddProcessor.propTypes = {
	processorTypes: PropTypes.array,
	onCreate: PropTypes.func,
	setNewProcessorType: PropTypes.func,
	value: PropTypes.string
};

AddProcessor.defaultProps = {
	processorTypes: [],
	value: ''
};
