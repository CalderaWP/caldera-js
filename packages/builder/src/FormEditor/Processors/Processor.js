import React, { Fragment, Component } from 'react';
import { TabPanel } from '@wordpress/components';
import {
	Row,
	Column,
	InputField,
	FORM_FIELDS_AUTO_COMPLETE_FIELD_TYPE_IDENTIFIER,
} from '@calderajs/components';
import { HorizontalForm } from '@calderajs/forms';
import PropTypes from 'prop-types';
import { AUTO_COMPLETE_FIELD_TYPE_IDENTIFIER } from '@calderajs/components';

const ProcessorLabel = ({ label }) => <div>{label}</div>;

const ProcessorLabelEdit = ({ label, onChange, id }) => (
	<InputField
		label={'Processor Label'}
		value={label}
		onChange={onChange}
		html5type={'text'}
		fieldId={`processor-label-${id}`}
	/>
);

const SettingsPanel = ({
	label,
	id,
	type,
	formProps,
	onEditLabel,
}) => {
	return (
		<div>
			<ProcessorLabel label={label ? label : type} />
			<ProcessorLabelEdit label={label} onChange={onEditLabel} id={id} />
			<HorizontalForm
				{...formProps}
				className={'caldera-processor-settings'}
			/>
		</div>
	);
};

export class Processor extends Component {
	state = {
		activeTab: 'settings',
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	tabs = [
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-processor-settings-tab-btn',
		},
		{
			name: 'conditionals',
			title: 'Conditionals',
			className: 'caldera-processor-conditionals-tab-btn',
		},
	];

	/**
	 * Prepare props for the processor editing form
	 *
	 * @return {Processor.props}
	 */
	formProps = () => {
		const { form } = this.props;
		let props = { ...this.props };
		props.fields.map(field => {
			if (
				FORM_FIELDS_AUTO_COMPLETE_FIELD_TYPE_IDENTIFIER ===
				field.fieldType
			) {
				field.form = form;
			}
			return field;
		});
		return props;
	};

	onEditLabel = label => {
		this.props.onChange({ label });
	};



	render() {
		console.log(this.props.conditonalPanel);
		const {
			onClose,
			onRemove,
			conditionalPanel,
		} = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={
							this.props.conditonalPanel
								? this.tabs
								: [this.tabs[0]]
						}
					>
						{() => {
							const { activeTab } = this.state;
							const tab = this.tabs.find(
								t => t.name === activeTab
							);
							const { name } = tab;
							if ('settings' === name) {
								return (
									<SettingsPanel
										{...this.props}
										formProps={this.formProps()}
										onEditLabel={this.onEditLabel}
									/>
								);
							}
							return (
								<div
									className={'caldera-processor-conditionals'}
								>
									{conditionalPanel}
								</div>
							);
						}}
					</TabPanel>
				</Row>
				<Row>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-close'}
							onClick={onClose}
						>
							Close
						</button>
					</Column>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-remove'}
							onClick={onRemove}
						>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		);
	}
}

Processor.propTypes = {
	...HorizontalForm.propTypes,
	label: PropTypes.string,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
	type: PropTypes.string,
	id: PropTypes.string,
	form: PropTypes.shape({
		id: PropTypes.string.isRequired,
		fields: PropTypes.array.isRequired,
	}).isRequired,
};

Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings',
	conditionalPanel: null,
};
