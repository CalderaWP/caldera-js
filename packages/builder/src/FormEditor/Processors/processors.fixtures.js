export const processorsCollection = [
	{
		"id": "season7Episode2",
		"label": "Test sending form data to test API",
		"type": "apiRequest",
		"config": {
			"requestURL": "https:\/\/something.com",
			"requestMethod": "POST",

		},
		fields: [
			{
				fieldId: 'requestURL',
				fieldType: 'input',
				html5type: 'url',
				required: true,
			},
			 {
			 	fieldId: 'requestMethod',
				fieldType: 'select',
				label: 'Request Method',
				default: 'GET',
				options: [
					{ value: "GET", label: 'GET' },
					{ value: "POST", label: 'POST' },
				]
			}

		]
	},
	{
		"id": "p2",
		"label": "Main MailChimp",
		"type": "mailchimp",
		"config": {
			"listId": "7",
		},
		fields: [
			{
				fieldId: 'listId',
				fieldType: 'input',
				html5type: 'text',
				required: true,
			}
		]
	}
];
