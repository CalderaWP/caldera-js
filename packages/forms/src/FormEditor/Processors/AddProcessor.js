import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Row, fieldAreaFactory} from "@calderawp/components";

export const AddProcessor = ({
								 setNewProcessorType,
								 processorTypes,
								 onCreate,
								 children,
								 value
							 }) => {
	const options = [
		{
			value: null,
			label: '--'
		}
	];

	processorTypes.map(processorType => {
		options.push({
			value: processorType.type,
			label: processorType.type
		});
	});


	const processorTypesField = {
		fieldType: 'select',
		label: 'Processor Type',
		fieldId: 'newProcessorType',
		required: true,
		options,
		onChange: setNewProcessorType,
		value,
	};


	return (
		<Fragment>
			{fieldAreaFactory(processorTypesField, setNewProcessorType)}
			<button onClick={onCreate}
					disabled={!value}
			>
				{children}
			</button>
		</Fragment>
	)
};

AddProcessor.propTypes = {
	processorTypes: PropTypes.array,
	onCreate: PropTypes.func,
	setNewProcessorType: PropTypes.func,
	value: PropTypes.string
};

AddProcessor.defaultProps = {
	processorTypes: [],
	value: ''
};
