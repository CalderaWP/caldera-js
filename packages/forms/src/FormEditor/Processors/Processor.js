import React, { Fragment, Component } from 'react';
import { HorizontalForm } from '../../HorizontalForm/HorizontalForm';
import { TabPanel } from '@wordpress/components';
import { Row, Column,FormFieldsAutoComplete } from"@calderawp/components";
import PropTypes from 'prop-types';

export class Processor extends Component {
	state = {
		activeTab: 'settings'
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	tabs = [
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-processor-settings-tab-btn'
		},
		{
			name: 'conditionals',
			title: 'Conditionals',
			className:
				'caldera-processor-conditionals-tab-btn'
		}
	];

	/**
	 * Prepare props for the processor editing form
	 *
	 * @return {Processor.props}
	 */
	formProps = () => {
		let {props} = this;
		const {fields,form} = this.props;
		const formFields = form.fields;
		fields.forEach( (field,i) => {
			const {fieldType} = field;
			if( FormFieldsAutoComplete.IDENTIFIER === fieldType ){
				props.fields[i] = {
					...field,
					form
				}
			}
		});
		return props;
	};

	render() {
		const { onClose, onRemove } = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={this.tabs}
					>
						{() => {
							const {activeTab} = this.state;
							const tab =  this.tabs.find( t => t.name === activeTab );
							const { name } = tab;
							if ('settings' === name) {
								return (
									<HorizontalForm
										{...this.formProps()}
										className={'caldera-processor-settings'}
									/>
								);
							}
							return (
								<div
									className={'caldera-processor-conditionals'}
								>
									Conditionals
								</div>
							);
						}}
					</TabPanel>
				</Row>
				<Row>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-close'}
							onClick={onClose}>
							Close
						</button>
					</Column>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-remove'}
							onClick={onRemove}>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		);
	}
}

//should have label!
Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
	type: PropTypes.string,
	id: PropTypes.string,
};

Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
