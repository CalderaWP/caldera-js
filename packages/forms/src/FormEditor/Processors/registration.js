

const defaultRegistry = [];

/**
 * Add a processor type to form editor
 *
 * @param {*} processorType Defintion of processor type
 * @param {Array} registry Optional. Collection to add to -- generally you leave this at default.
 * @return {Array}
 */
export const registerProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		return [ processorType,...registry];
	}

	return registry;

};

/**
 * Get a processor type
 *
 * @param {string} processorType
 * @param {Array} registry
 * @return {*}
 */
export const unregisterProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		processorType = processorType.type;
	}
	const index = registry.findIndex( p => p.type === processorType );
	if( 0<= index ){
		registry.splice(index,1);
		return [ ...registry];
	}
	return registry;
};

/**
 * Get all processors from collection
 *
 * @param registry
 * @return {Array}
 */
export const getProcessors = (registry = defaultRegistry) => {
	return registry;
};
