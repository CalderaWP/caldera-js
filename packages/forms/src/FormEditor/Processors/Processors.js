import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";
import {Row, fieldAreaFactory} from "@calderawp/components";
import {processorTypesPropType} from './processorTypesPropType';
import processorFactory from './processorTypes/processorFactory';
import {AddProcessor} from './AddProcessor';


export class Processors extends Component {

	state = {
		activeProcessorId: '',
		newProcessorType: ''
	};

	/**
	 * Set which processor is active
	 * @param activeProcessorId
	 */
	setActive = (activeProcessorId) => {
		this.setState({activeProcessorId});
	};

	/**
	 * Check if a processor is active by ID
	 * @param processorId
	 * @return {string|boolean}
	 */
	isActiveProcessor = (processorId) => {
		const {activeProcessorId} = this.state;
		return activeProcessorId && processorId === activeProcessorId;
	};

	/**
	 * When a change in value for a processor is received, send all processors up.
	 *
	 * @param processorId
	 * @param updatedProcessor
	 */
	handleProcessorChange = (processorId, updatedProcessor) => {
		const {processors, updateProcessors} = this.props;
		let processorIndex = processors.findIndex(processor => {
			return processorId === processor.id
		});
		if (-1 !== processorIndex) {
			let processor = processors[processorIndex];
			let update = [
				...processors.splice(processorIndex - 1, 1),
				{
					...processor,
					...updatedProcessor
				}
			];
			updateProcessors(update);
		}


	};

	/**
	 * When a processor is removed, send updated list to change handler.
	 *
	 * @param processorId
	 */
	handleRemoveProcessor = (processorId) => {
		const {processors, updateProcessors} = this.props;
		updateProcessors([...processors.filter(processor => {
			return processorId !== processor.id
		})]);
	};

	/**
	 * When a processor is added, create send update list to change handler

	 */
	handleCreateProcessor = () => {
		const {processors, updateProcessors} = this.props;
		const {activeProcessor,newProcessorType} = this.state;
		const newProcessor = processorFactory(newProcessorType);
		const {id} = newProcessor;
		updateProcessors([...processors, newProcessor]);
		this.setState({activeProcessorId:id,newProcessorType: ''});
	};

	/**
	 * Set the type of the next processor to be created
	 *
	 * @param newProcessorType
	 * @return {*}
	 */
	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	/** @inheritDoc **/
	render() {
		const {activeProcessor,newProcessorType} = this.state;
		const {processors,processorTypes,form} = this.props;
		return (
			<div>
				<div>
					{processors.map(processor => {
						const {
							id,
							fields,
							config,
							type,
							label,
							conditionals
						} = processor;
						if (this.isActiveProcessor(id)) {
							return (
								<Fragment key={id}>
									<Processor
										form={form}
										className={`caldera-forms-active-processor-${id}`}
										fields={fields}
										conditionals={conditionals}
										initialValues={config}
										label={label}
										type={type}
										onChange={(fieldValues) => this.handleProcessorChange(id, fieldValues)}
										onRemove={() => this.handleRemoveProcessor(id)}
										onClose={() => this.setActive('')}
									/>
								</Fragment>
							)
						}
						return (
							<Row key={id}>
								<button
									className={`caldera-forms-choose-processor caldera-forms-choose-processor-${id}`}
									onClick={() => this.setActive(id)}
								>
									{label ? label : type}
								</button>
							</Row>

						)
					})}
				</div>
				<div>
					<AddProcessor
						setNewProcessorType={this.setNewProcessorType}
						processorTypes={processorTypes}
						value={newProcessorType}
						onCreate={this.handleCreateProcessor}
					>
						Add {! newProcessorType ? '' : newProcessorType } Processor
					</AddProcessor>
				</div>
			</div>
		);
	}
}

export const processorsCollectionPropType = PropTypes.arrayOf(
	PropTypes.shape({
		id: PropTypes.string,
		label: PropTypes.string,
		type: PropTypes.string,
		fields: PropTypes.array,
		config: PropTypes.object
	})
);

Processors.propTypes = {
	processorTypes: processorTypesPropType,
	processors: processorsCollectionPropType,
	updateProcessors: PropTypes.func,
	form: PropTypes.object,
	formFields: PropTypes.array
};

Processors.defaultProps = {
	processors: [],
	processorTypes: []
};
