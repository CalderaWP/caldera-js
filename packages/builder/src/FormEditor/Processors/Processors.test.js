import React, { Component } from 'react';

import { Processors } from './Processors';
import { mount } from 'enzyme';
import { processorsCollection } from './processors.fixtures';
import { render,fireEvent,cleanup } from 'react-testing-library';

describe('Processors', () => {
	afterEach(cleanup);
	const processors = [
		{ id: 'p1', type: 'Redirect', fields: [] },
		{ id: 'p2', type: 'Redirect', fields: [] },
		{ id: 'p3', type: 'Redirect', fields: [] },
	];

	const form = {};
	const formFields = [];
	let updateProcessors;
	beforeEach(() => {
		updateProcessors = jest.fn();
	});

	it.skip('Should update active', () => {
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().setActive('p2');
		expect(component.state('activeProcessorId')).toEqual('p2');
	});

	it.skip('Can remove a processor', () => {
		let values = {};
		const updateProcessors = updateValues => {
			values = updateValues;
		};
		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().handleRemoveProcessor('p2');

		expect(values.length).toEqual(2);
	});

	it.skip('Can update a processor type', () => {
		const processors = [
			{ id: 'p1', type: 'Redirect', fields: [] },
			{ id: 'p2', type: 'Redirect', fields: [], label: 'fLabel' },
		];
		let values = {};
		const updateProcessors = updateValues => {
			values = updateValues;
		};

		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.instance().handleProcessorChange('p2', {
			type: 'strident',
		});
		expect(values.find(p => 'p2' === p.id).label).toEqual('fLabel');
		expect(values.find(p => 'p2' === p.id).type).toEqual('strident');
	});

	it.skip('Does not call processor update if processor id is invalid', () => {
		const processors = [
			{ id: 'p1', type: 'Redirect', fields: [] },
			{ id: 'p2', type: 'Redirect', fields: [], label: 'fLabel' },
		];

		const component = mount(
			<Processors
				processors={processors}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.instance().handleProcessorChange('p166161111', {
			type: 'strident',
		});
		expect(updateProcessors.mock.calls.length).toEqual(0);
	});
});

describe('Processors collection in processors UI', () => {
	afterEach(cleanup);

	const form = {};
	const formFields = [];
	let updateProcessors;
	beforeEach(() => {
		updateProcessors = jest.fn();
	});

	it.skip('Renders wit.skiph a list of processors', () => {
		const component = render(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		expect(component).toMatchSnapshot();
	});

	it.skip('should open processor edit.skipor', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		component.find('.caldera-forms-choose-processor-p2').simulate('click');
		expect(component.state('activeProcessorId')).toEqual('p2');
	});

	it.skip('Calls the processor update callback when a processor changes values', () => {
		const updateProcessors = jest.fn();
		const {container} = render(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = { target: { value: 'six' } };

		fireEvent.click(container.querySelector('.caldera-forms-choose-processor-p2') );

		  fireEvent.change(
			  container.querySelector('.caldera-forms-active-processor-p2 input[type="text"]'),
			event
			);
		  expect(updateProcessors).toHaveBeenCalledTimes(1);

	});

	it.skip('Calls the processor update callback when a processor is removed', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = { target: { value: 'six' } };

		component.find('.caldera-forms-choose-processor-p2').simulate('click');
		component.find('.caldera-processor-remove').simulate('click');
		expect(updateProcessors.mock.calls.length).toEqual(1);
	});

	it.skip('Closes the processor', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);

		const event = { target: { value: 'six' } };

		component.find('.caldera-forms-choose-processor-p2').simulate('click');
		component.find('.caldera-processor-close').simulate('click');
		expect(component.state('activeProcessorId')).toEqual('');
	});

	it.skip('sets newProcessorType', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.instance().setNewProcessorType('redirect');
		expect(component.state('newProcessorType')).toBe('redirect');
	});

	it.skip('Adds new processor when created', () => {
		const component = mount(
			<Processors
				processors={processorsCollection}
				form={form}
				formFields={formFields}
				updateProcessors={updateProcessors}
			/>
		);
		component.setState({ newProcessorType: 'autoResponder' });
		component.instance().handleCreateProcessor();
		expect(updateProcessors.mock.calls.length).toEqual(1);
		expect(
			updateProcessors.mock.calls[0][0].find(
				p => p.type === 'autoResponder'
			).type
		).toEqual('autoResponder');
	});
});
