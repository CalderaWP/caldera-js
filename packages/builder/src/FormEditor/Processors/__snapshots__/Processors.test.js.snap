// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processors collection in processors UI Renders with a list of processors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-atomic="true"
      aria-live="polite"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-polite"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div
      aria-atomic="true"
      aria-live="assertive"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-assertive"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div>
      <div>
        <div>
          <div
            class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
          >
            <button
              class="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
            >
              Test sending form data to test API
            </button>
          </div>
          <div
            class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
          >
            <button
              class="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
            >
              Main MailChimp
            </button>
          </div>
        </div>
        <div>
          <div
            class="caldera-field-group caldera-field-area-select is-required form-group"
          >
            <label
              class="form-label"
              for="newProcessorType"
            >
              Processor Type
               
              <span>
                *
              </span>
            </label>
            <select
              class="caldera-field caldera-field-select form-control"
              id="newProcessorType"
              required=""
            >
              <option>
                --
              </option>
              <option>
                --
              </option>
            </select>
            
          </div>
          <button
            disabled=""
          >
            Add 
            
             Processor
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div
          class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
        >
          <button
            class="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
          >
            Test sending form data to test API
          </button>
        </div>
        <div
          class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
        >
          <button
            class="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
          >
            Main MailChimp
          </button>
        </div>
      </div>
      <div>
        <div
          class="caldera-field-group caldera-field-area-select is-required form-group"
        >
          <label
            class="form-label"
            for="newProcessorType"
          >
            Processor Type
             
            <span>
              *
            </span>
          </label>
          <select
            class="caldera-field caldera-field-select form-control"
            id="newProcessorType"
            required=""
          >
            <option>
              --
            </option>
            <option>
              --
            </option>
          </select>
          
        </div>
        <button
          disabled=""
        >
          Add 
          
           Processor
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
